### PR TITLE
Dev 3 added solution

### PR DESCRIPTION
**Pull request by Nokafan**
I used try with resources for all autoclosable it is too many information about this, it is good practice to close all. Some db could not close result set. But if it has to be removed I'll do it :)

### CODE REVIEW

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+public class OrderDaoJdbcImpl implements OrderDao {
+
+    @Override
+    public List<Order> getUserOrders(Long id) {
+        String query = "SELECT * FROM orders WHERE user_id = ? AND order_deleted = FALSE;";
+        List<Order> ordersList = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+                preparedStatement.setLong(1, id);
+                try (ResultSet resultSet = preparedStatement.executeQuery()) {
```
Comment (TimoxaV) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: resultSet закроется вместе с Statement -можно без try with resources с ним работать

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+                        order.setId(resultSet.getLong("order_id"));
+                        order.setUserId(resultSet.getLong("user_id"));
+                    }
+                }
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't get order id=" + id, e);
+        }
+        order.setProducts(insertProductsToOrder(order));
+        return Optional.of(order);
```
Comment (TimoxaV) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:79]: Что если ордер с заданным id не найден?

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+@Dao
+public class OrderDaoJdbcImpl implements OrderDao {
+
+    @Override
+    public List<Order> getUserOrders(Long id) {
+        String query = "SELECT * FROM orders WHERE user_id = ? AND order_deleted = FALSE;";
+        List<Order> ordersList = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
```
Comment (TimoxaV) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:26]: Statement можно поместить в одном блоке с Connection. Тогда можно обойтись одним try with resources

**Review by TimoxaV — COMMENTED** [review]
Замечания к OrderDaoJdbc актуальны и к ShoppingCartDaoJdbc, и к UserDaoJdbc

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 2 above)
Comment (Nokafan) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:79]: SQL exception з id, що прилетіло в тіло методу.

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 1 above)
Comment (Nokafan) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: https://docs.oracle.com/javadb/10.8.3.0/devguide/cdevconcepts839085.html
https://stackoverflow.com/questions/41555143/java-sql-is-it-mandatory-to-close-a-resultset-statement-connection
https://www.mysqltutorial.org/querying-data-from-mysql-using-jdbc/

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 3 above)
Comment (Nokafan) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:26]: Да но тоді у мене чекстайл сходить з ума і видає 35 errors ((((((

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 1 above)
Comment (Nokafan) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: Видалив трай для резалт сета

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Dao
+public class OrderDaoJdbcImpl implements OrderDao {
+    @Override
+    public List<Order> getUserOrders(Long id) {
+        String query = "SELECT * FROM orders WHERE user_id = ? AND order_deleted = FALSE;";
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:23]:
```suggestion
        String query = "SELECT * FROM orders WHERE user_id = ? AND = deleted = FALSE;";
```
it's field in table ordes— no need to add order to column name

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+                         connection.prepareStatement(deleteProducts)) {
+                preparedStatement.setLong(1, id);
+                return preparedStatement.executeUpdate();
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't delete products from order id=" + id, e);
+        }
+    }
+
+    private List<Product> insertProductsToOrder(Order order) {
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]:
```suggestion
    private List<Product> getOrderProducts(Order order) {
```

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+                preparedStatement.setLong(1, order.getId());
+                ResultSet resultSet = preparedStatement.executeQuery();
+                while (resultSet.next()) {
+                    productList.add(getProductFromResultSet(resultSet));
+                }
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Cuoldn't set products to order from DB ", e);
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: let's add order id to exception message

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+                ResultSet generatedKeys = preparedStatement.getGeneratedKeys();
+                if (generatedKeys.next()) {
+                    cart.setId(generatedKeys.getLong(1));
+                }
+
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("ShoppingCart for userId=" + cart.getUserId()
+                    + " wasn't created.", e);
+        }
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java:58]: let's also insert products from shopping cart here

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        deleteProductsFromCart(cart);
+        String query = "INSERT INTO shopping_carts_products (cart_id, product_id) VALUES (?, ?);";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            for (Product product : cart.getProducts()) {
+                try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+                    preparedStatement.setLong(1, cart.getId());
+                    preparedStatement.setLong(2, product.getId());
+                    preparedStatement.executeUpdate();
+                }
+            }
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java:114]: it's better to move this  to separate method and use here and in create()

**Review by dmytro-hryhoruk — APPROVED** [review]

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 4 above)
Comment (Nokafan) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:23]: It's the name of column order_deleted, not name of table.

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 5 above)
Comment (Nokafan) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: Fixed

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java (hunk 7 above)
Comment (Nokafan) [reply src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java:58]: Fixed.

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java (hunk 8 above)
Comment (Nokafan) [reply src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java:114]: But when we are creating the cart we do not insert any product into it. We only create an empty cart and we do set the id to cart that's it.

Comment (Nokafan) [reply src/main/java/com/internet/shop/dao/jdbc/ShoppingCartJdbcImpl.java:114]: By the way, I pushed HW on Friday morning before some people :) and their hw was approved, it is unfair :)

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 6 above)
Comment (Nokafan) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java]: Fixed.

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 4 above)
Comment (Nokafan) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:23]: image

**Pull request merged**